Node builds broken on master

## Hello!
<!-- Please leave this section as-is, it's designed to help others in the community know how to interact with our GitHub issues. -->

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

```
cd sdk/nodejs && make ensure build
```

```
node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18 - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
  The types returned by 'end(...)' are incompatible between these types.
    Type 'void' is not assignable to type 'this'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: string, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.

64     end(metadata?: any): void;
       ~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: string | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.

77     end(metadata?: any): void;
       ~~~


Found 3 errors.
```

https://github.com/pulumi/pulumi/runs/4696145791?check_suite_focus=true#step:18:30

<!-- Please provide a general summary of the issue, including what you're trying to accomplish. -->

### Steps to reproduce
<!-- Provide a link to a live example or an unambiguous set of steps to reproduce this bug. Tell us what you expected to happen, and what happened instead. Please include code to reproduce, if relevant. -->

1.
2.
3.
4.

Expected: <!-- What did you expect to happen? -->
Actual: <!-- What happens instead? -->

--- a/sdk/nodejs/package.json
+++ b/sdk/nodejs/package.json
@@ -29,7 +29,7 @@
         "@types/js-yaml": "^3.12.5",
         "@types/minimist": "^1.2.0",
         "@types/mocha": "^2.2.42",
-        "@types/node": "^12.0.0",
+        "@types/node": "14.18.3",
         "@types/normalize-package-data": "^2.4.0",
         "@types/read-package-tree": "^5.2.0",
         "@types/semver": "^5.5.0",
@@ -50,5 +50,8 @@
     },
     "mocha": {
         "require": ["ts-node/register", "source-map-support/register"]
-    }
+    },
+    "//": [
+       "NOTE: @types/node is pinned due to grpc/grpc-node#2002"
+    ]
 }

--- a/sdk/nodejs/runtime/closure/createClosure.ts
+++ b/sdk/nodejs/runtime/closure/createClosure.ts
@@ -1358,7 +1358,8 @@ async function findNormalizedModuleNameAsync(obj: any): Promise<string | undefin
     // don't pre-compute this because the require cache will get populated
     // dynamically during execution.
     for (const path of Object.keys(require.cache)) {
-        if (require.cache[path].exports === obj) {
+        const c = require.cache[path];
+        if (c !== undefined && c.exports === obj) {
             // Rewrite the path to be a local module reference relative to the current working
             // directory.
             const modPath = upath.relative(process.cwd(), path);

